gcm/gcm.py

```diff
@@ -96,13 +92,65 @@ def urlencode_utf8(params):
     return '&'.join(params)
 
 
+class Payload(object):
+    """
+    Base Payload class which prepares data for HTTP requests
+    """
+
+    # TTL in seconds
+    GCM_TTL = 2419200
+
+    def __init__(self, **kwargs):
+        self.validate(kwargs)
+        self.__dict__.update(**kwargs)
+
+    def validate(self, options):
+        """
+        Allow adding validation on each payload key
+        by defining `validate_{key_name}`
+        """
+        for key, value in options.items():
+            validate_method = getattr(self, 'validate_%s' % key, None)
+            if validate_method:
+                validate_method(value)
+
+    def validate_time_to_live(self, value):
+        if not (0 <= value <= self.GCM_TTL):
+            raise GCMInvalidTtlException("Invalid time to live value")
+
+    @property
+    def body(self):
+        raise NotImplementedError
+
+
+class PlaintextPayload(Payload):
+
+    @property
+    def body(self):
+        # Safeguard for backwards compatibility
+        if 'registration_id' not in self.__dict__:
+            self.__dict__['registration_id'] = self.__dict__.pop(
+                'registration_ids', None
+            )
+        # Inline data for for plaintext request
+        data = self.__dict__.pop('data')
+        for key, value in data.items():
+            self.__dict__['data.%s' % key] = value
```

Comment:
Adding 'data.' to each key inside the data dict adds up to the 4KB limit if you have many keys. For some people this could be an issue if they send messages that are close to the limit and due to the key length they can't send as much data as they want.